ffai/core/forward_model.py

```diff
@@ -0,0 +1,359 @@
+
+        set.add(self, value)
+
+    def clear(self):
+        if self.logger_initialized():
+            log_entry = CallableStep(self, forward_func=set.clear, forward_args=(),
+                                     backward_func=set.update, backward_args=(copy(self),))
+            self.log_this(log_entry)
+
+        set.clear(self)
+
+    def pop(self):
+        raise NotImplementedError()
+
+    def remove(self, value):
+        raise NotImplementedError()
+
+    def __reduce__(self):
+        func = LoggedSet.init_LoggedSet
+        values = set()
+        values.update(self)
+        return func, (values, self._logger)
+
+    def init_LoggedSet(values, logger):  # Static method
+        logged_set = LoggedSet(values)
+        object.__setattr__(logged_set, "_logger", logger)
+        return logged_set
+
+
+class LoggedDict(dict, LoggedState):
+    def __init__(self, value):
+        super().__init__(value)
+        LoggedState.__init__(self)
+
+    def set_logger(self, logger):
+        if self.logger_initialized():
+            return
+
+        LoggedState.set_logger(self, logger)
+        for key in self:
+            new_value = add_logging(self[key], logger)
+            dict.__setitem__(self, key, new_value)
+
+    def __setitem__(self, key, value):
+        if key in self.keys():
+            raise NotImplementedError()
+        else:
+            if self.logger_initialized():
+                value = add_logging(value, self._logger)
+            super().__setitem__(key, value)
+
+    def pop(self, key):
+        raise NotImplementedError()
+
+    def clear(self):
+        raise NotImplementedError()
+
+    def popitem(self):
+        raise NotImplementedError()
+
+    def __reduce__(self):
+        func = LoggedDict.init_LoggedDict
+        values = {}
+        values.update(self)
+        return func, (values, self._logger)
+
+    def init_LoggedDict(values, logger):  # Static method
+        logged_dict = LoggedDict(values)
+        object.__setattr__(logged_dict, "_logger", logger)
+        return logged_dict
+
+
+def is_immutable(obj):
+    return type(obj) in immutable_types or isinstance(obj, Enum) or isinstance(obj, Immutable)
+
+
+replacement_type = [(list, LoggedList), (dict, LoggedDict), (set, LoggedSet)]
+immutable_types = {int, float, str, tuple, bool, range, type(None)}
+
+
+def add_logging(value, logger):
+    if is_immutable(value):
+        return value
+
+    if isinstance(value, LoggedState):
+        if not value.logger_initialized():
+            value.set_logger(logger)
+        return value
+
+    new_types = [t[1] for t in replacement_type if type(value) == t[0]]
+    if len(new_types) == 1:
+        new_type = new_types.pop()
+        new_value = new_type(value)
+        new_value.set_logger(logger)
+        return new_value
+    else:
+        raise AttributeError(f"Unable to add logging to {value}")
+
+
+class Immutable:
+    def __setattr__(self, key, value):
+        if hasattr(self, key):
```

Comment:
Immutable's __setattr__ is responsible for a lot of computation time according to the cProfiler.